src/Ardalis.Extensions/Tasks/WhenAll.cs

```diff
@@ -0,0 +1,388 @@
+using System;
+using System.Threading.Tasks;
+
+namespace Ardalis.Extensions.Tasks;
+
+public static partial class TasksExtensions
+{
+
+  /// <summary>
+  /// Creates a task that will complete when all of the supplied tasks have completed.
+  /// </summary>
+  /// <param name="tasks">The tasks to wait on for completion.</param>
+  /// <returns>A task that represents the completion of all of the supplied tasks.</returns>
+  /// <remarks>
+  /// <para>
+  /// If any of the supplied tasks completes in a faulted state, the returned task will also complete in a Faulted state,
+  /// where its exceptions will contain the aggregation of the set of unwrapped exceptions from each of the supplied tasks.
+  /// </para>
+  /// <para>
+  /// If none of the supplied tasks faulted but at least one of them was canceled, the returned task will end in the Canceled state.
+  /// </para>
+  /// <para>
+  /// If none of the tasks faulted and none of the tasks were canceled, the resulting task will end in the RanToCompletion state.
+  /// </para>
+  /// </remarks>
+  /// <exception cref="System.ArgumentException">
+  /// The <paramref name="tasks"/> tuple contained a null task.
+  /// </exception>
+  public static async Task<(T1, T2)> WhenAll<T1, T2>(this (Task<T1>, Task<T2>) tasks)
+  {
+    var task = Task.WhenAll(tasks.Item1, tasks.Item2);
+    try
+    {
+      await task;
+    }
+    catch (Exception e)
+    {
+      if (task.Exception != null)
+      {
+        throw task.Exception;
+      }
+
+      throw e;
```

Comment:
As the warning suggests here, you just want to `throw;` as opposed to `throw e;`